1. WA function to take a number as parameter and  print True if given number is even else print False

In [1]:
def even(e):
    if(e % 2==0):
        return True
    else:
        return False
print(even(10))        

True


2. WA function to take list as parameter and print all numbers of given list which are divisible by 11

In [6]:
def elev(el:list):
    for i in el:
        if(i % 11==0):
            print(i)
    return        
el = [11,2,5,66,3,99,7,33]
elev(el)

11
66
99
33


3. WA function to take string as parameter
Check if string is palindrome or not
If palindrome return 1
if not palindrome return 0

In [16]:
def palin(st):
    if(st == st[::-1]):
        return True
    else:
        return False
        
palin("harah")            
            
    
    

True

4. WA function to check if number is prime or not. Return True if number is prime and false if number not prime.

In [1]:
def prime(n):
    idx = 2
    while(idx < int(n**0.5) + 1):
        if(n%idx == 0):
            return False
        idx += 1
    else:
        return True
print(prime(9))        

False


Q WA function to take a string as parameter. Print its alternate characters starting from 2nd character

In [3]:
def alt(s):
    print(s[1::2])

s = "elephant"
alt(s)    

lpat


Q WA function to take dictionary as parameter. Print all keys of the dictionary

In [8]:
def dict(dic):
     print(dic.keys())

dic = {1:10,2:20,3:30}
dict(dic)


dict_keys([1, 2, 3])


Q WA function to take a dictionary as parameter. Print value of key "name" if key "rollno" is having value 100

In [14]:
def std(d):
    if(d['rollno'] == 100):
        print(d['name'])
    else:
        print("not found")
d = {'name': 'harsh','rollno': 100}
std(d)

harsh


Q WA function to take list of strings as parameter and print all strings which have character 's' 2 times

In [17]:
def s2(ls):
    for i in ls:
        if(i.count('s') == 2):
            print(i)
ls = ["kassd",'hufsca','ershfs']
s2(ls)

kassd
ershfs


Q WA function to take list of dictionaries. Print all dictionaries which contain key "id".

In [19]:
def idd(ld):
    for i in ld:
        if('id' in i.keys()):
            print(i)
ld = [{'name': 'harsh','id': 100},{'id': 101},{'name':'adi'}]
idd(ld)


{'name': 'harsh', 'id': 100}
{'id': 101}


## function Practice

1. Write a Python function to find the maximum of three numbers.

In [23]:
def max3(a,b,c):
    if(a > b):
        if(a > c):
            print(a,' is greater')
        else:
            print(c,' is grater')
    else:
        if(b>c):
            print(b, ' is greater')
        else:
            print(c, ' is greater')
    return

a=10
b=20
c=15
max3(a,b,c)
    

20  is greater


2. Write a Python function to sum all the numbers in a list. 
Sample List : (8, 2, 3, 0, 7)
Expected Output : 20

In [26]:
def summ(ls):
    x=0
    for i in ls:
        x += i
    print("sum:", x)
ls = [8, 2, 3, 0, 7]
summ(ls)

sum: 20


3. Write a Python function to multiply all the numbers in a list. 
Sample List : (8, 2, 3, -1, 7)
Expected Output : -336

In [28]:
def mul(m):
    mul=1
    for i in m:
        mul *= i
    print("multiplication =",mul)

m = [8, 2, 3, -1, 7]
mul(m)

multiplication = -336


4. Write a Python program to reverse a string. 
Sample String : "1234abcd"
Expected Output : "dcba4321"

In [30]:
def rev(g):
    print(g[::-1])

g = "1234abcd"
rev(g)

dcba4321


5. Write a Python function to calculate the factorial of a number (a non-negative integer). The function accepts the number as an argument.

In [32]:
def fact(n:int):
    p=1
    for i in range(2,n+1):
        p = p * i
    return p
print(fact(5))

120


6. Write a Python function to check whether a number falls within a given range. 

7. Write a Python function that accepts a string and counts the number of upper and lower case letters. 
Sample String : 'The quick Brow Fox'
Expected Output :
No. of Upper case characters : 3
No. of Lower case Characters : 12

In [38]:
def ul(y):
    u=0
    l=0
    for i in y:
        if(i.isupper()):
            u+=1
        if(i.islower()):
            l+=1
    print("Lower", l,"upper:",u)        

t = 'The quick Brow Fox'
ul(t)

Lower 12 upper: 3


8. Write a Python function that takes a list and returns a new list with distinct elements from the first list. 
Sample List : [1,2,3,3,3,3,4,5]
Unique List : [1, 2, 3, 4, 5]

In [41]:
def unq(sl):
    ul =[]
    for i in sl:
        if(i not in ul):
            ul.append(i)
    print(ul)        
sl= [1,2,3,3,3,3,4,5]
unq(sl)


[1, 2, 3, 4, 5]


9. Write a Python function that takes a number as a parameter and checks whether the number is prime or not. 
Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [42]:
def prime(n):
    idx = 2
    while(idx < int(n**0.5) + 1):
        if(n%idx == 0):
            return False
        idx += 1
    else:
        return True
print(prime(9)) 

False


10. Write a Python program to print the even numbers from a given list. 
Sample List : [1, 2, 3, 4, 5, 6, 7, 8, 9]
Expected Result : [2, 4, 6, 8]

In [45]:
def even(L):
    e=[elem for elem in L if elem % 2== 0]
    print(e)
L = [1, 2, 3, 4, 5, 6, 7, 8, 9]
even(L)

[2, 4, 6, 8]


11. Write a Python function to check whether a number is "Perfect" or not. 
According to Wikipedia : In number theory, a perfect number is a positive integer that is equal to the sum of its proper positive divisors, that is, the sum of its positive divisors excluding the number itself (also known as its aliquot sum). Equivalently, a perfect number is a number that is half the sum of all of its positive divisors (including itself).
Example : The first perfect number is 6, because 1, 2, and 3 are its proper positive divisors, and 1 + 2 + 3 = 6. Equivalently, the number 6 is equal to half the sum of all its positive divisors: ( 1 + 2 + 3 + 6 ) / 2 = 6. The next perfect number is 28 = 1 + 2 + 4 + 7 + 14. This is followed by the perfect numbers 496 and 8128.

In [56]:
def perfect(n):
    sum =0
    for i in range(1,n):
        if(n % i == 0):
            print(i)
            sum += i
    if(sum == n and (sum+n)/2 == n):
        print("Perfect!!!")
    else:
        print("not perfect")

n = 28
perfect(n)

1
2
4
7
14
Perfect!!!


12. Write a Python function that checks whether a passed string is a palindrome or not. 
Note: A palindrome is a word, phrase, or sequence that reads the same backward as forward, e.g., madam or nurses run.

In [60]:
def palin(st):
    if(st == st[::-1]):
        return True
    else:
        return False
        
palin("harah")            

True

13. Write a Python function that prints out the first n rows of Pascal's triangle. 
Note : Pascal's triangle is an arithmetic and geometric figure first imagined by Blaise Pascal.

Sample Pascal's triangle :
Pascal's triangle

Each number is the two numbers above it added together 

In [5]:
from math import factorial
n = 5
for i in range(n):
    for j in range(n-i):
        print(end=" ")
        
    for j in range(i+1):
        # nCr = n!/((n-r)!*r!)
        print(factorial(i)//(factorial(j)*factorial(i-j)),end=" ")

    print()

     1 
    1 1 
   1 2 1 
  1 3 3 1 
 1 4 6 4 1 


14. Write a Python function to check whether a string is a pangram or not. 
Note : Pangrams are words or sentences containing every letter of the alphabet at least once.
For example : "The quick brown fox jumps over the lazy dog"

In [19]:
st = "The quick brown fox jumps over the lazy do"
st = st.lower()
ls = ['']*26
count =0
for i in st:
    if(i.isalpha()):
        idx = ord(i) - ord('a')
        ls[idx] = i
print(ls)
for i in ls:
    if(i == ""):
        count += 1
if(count >0):
    print("no pangrams")
else:
    print("Pangrams")

['a', 'b', 'c', 'd', 'e', 'f', '', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
no pangrams


15. Write a Python program that accepts a hyphen-separated sequence of words as input and prints the words in a hyphen-separated sequence after sorting them alphabetically. 
Sample Items : green-red-yellow-black-white
Expected Result : black-green-red-white-yellow

In [6]:
def hypsep(em):
    l =em.split("-")
    l.sort()
    print("-".join(l))
    
em = "green-red-yellow-black-white"
hypsep(em)
    

black-green-red-white-yellow


16. Write a Python function to create and print a list where the values are the squares of numbers between 1 and 30 (both included).

In [8]:
def cl():
    lt =[]
    for i in range(1,31):
        lt.append(i*i)
    print(lt)

cl()

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900]


In [9]:
def facto(n):
    if(n ==1):
        return n
    else:
        return n*facto(n-1)
 
facto(5)



120

[]

## Lambda

Q Sort all the numbers in a list in descending order using lambda function
sorted(l1, key=(lambda x:-x))

In [2]:
l1 = [1,7,3,8,2,5,9]
print(sorted(l1, key=(lambda x:-x)))

[9, 8, 7, 5, 3, 2, 1]


Q Sort all the characters of given string using lambda function

In [4]:
st = "harsh"
print(sorted(st),lambda x :x)

['a', 'h', 'h', 'r', 's'] <function <lambda> at 0x000001A2ED7A6520>


Q Given list of strings, sort all the strings by last character of that string. . Use lambda function and normal function both.

In [6]:
st = ["hei","harsh","happy","sad"]
print(sorted(st,key = lambda x : x[-1]))

['sad', 'harsh', 'hei', 'happy']


Q Given list of employees. This list may contain repetitions. Find all unique employee names and print them as per order of second character in that name. Use lambda function and normal function both.

In [11]:
emp = ["ram","sumit","aryan","rohit","aryan"]
emp2 = list(set(emp))
print(sorted(emp2,key = lambda x:x[1]))

['ram', 'rohit', 'aryan', 'sumit']


Q. Take Employee id and Employee name of 5 employees from user. Store it in dictionary. Print all employees in increasing order of employee Id . Also print all employees in alphabetical order by name.

In [29]:

emp ={1:"harsh",2:"yash",4:"deepa", 3:"sumit"}
'''for i in range(5):
    n = input("enter name ")
    r = int(input("enter roll no: "))
    emp[r] = n'''
print(emp)
a= dict(sorted(emp.items(),key = lambda k: k ))
print(a)
#print(sorted(emp,key = lambda v: ))
b = dict(sorted(emp.items(),key = lambda x: x[1]))
print(b)

{1: 'harsh', 2: 'yash', 4: 'deepa', 3: 'sumit'}
{1: 'harsh', 2: 'yash', 3: 'sumit', 4: 'deepa'}
{4: 'deepa', 1: 'harsh', 3: 'sumit', 2: 'yash'}


## Map

Q. Take comma separated numbers as input from the user. Split it in list of strings. Now convert every string in this list to float using map function

In [34]:
num = input("enter numbers")
sp = num.split(",")
res = map(lambda x: float(x), sp)
list(res)

enter numbers 1,2,3,4,5,6,7


[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0]

Q Given list of floating point numbers. Convert every number into string and then join all the numbers such that they are separated by pipe(|)

In [44]:
fl = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0]
res = map(str,fl)
"|".join(res)
    


'1.0|2.0|3.0|4.0|5.0|6.0|7.0'

Q Given list of integers. Use map function to calculate square of all and store in list. Then print sum of all square values.

In [47]:
ls = [1,2,3,4,5,6,7]
sq = []
res = map(lambda x : x*x, ls)
sm =0
for i in res:
    sq.append(i)
    sm += i
print(sq)
print(sm)
    

[1, 4, 9, 16, 25, 36, 49]
140


## function practice lambda"

3. Write a Python program to sort a list of tuples using Lambda.
Original list of tuples:
[('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)]
Sorting the List of Tuples:
[('Social sciences', 82), ('English', 88), ('Science', 90), ('Maths', 97)]

In [48]:
marks = [('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)]
print(sorted(marks, key = lambda x : x[1]))

[('Social sciences', 82), ('English', 88), ('Science', 90), ('Maths', 97)]


5. Write a Python program to filter a list of integers using Lambda.
Original list of integers:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Even numbers from the said list:
[2, 4, 6, 8, 10]
Odd numbers from the said list:
[1, 3, 5, 7, 9]

In [50]:
it = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
le = list(filter(lambda x : x %2 == 0,it))
lo = list(filter(lambda x : x %2 != 0,it))
print(le,lo,sep =" ")

[2, 4, 6, 8, 10] [1, 3, 5, 7, 9]


6. Write a Python program to square and cube every number in a given list of integers using Lambda.
Original list of integers:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Square every number of the said list:
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
Cube every number of the said list:
[1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]

In [56]:
num = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
#sqr = lambda x : print(x**2)
sq= lambda x: print(x**2)
for i in num:
    sq(i)

1
4
9
16
25
36
49
64
81
100


7. Write a Python program to find if a given string starts with a given character using Lambda.
Sample Output:
True
False

TypeError: 'function' object is not an iterator